test/service_manager_test.dart

```diff
@@ -165,6 +166,154 @@ void main() {
 
     // TODO(kenzie): add hot restart test case.
   }, tags: 'useFlutterSdk');
+
+  group('serviceManagerTests - restoring device-enabled extension:', () {
+    FlutterRunTestDriver _flutter;
+    String _flutterIsolateId;
+    VmServiceWrapper service;
+
+    setUp(() async {
+      setGlobal(ServiceConnectionManager, new ServiceConnectionManager());
+
+      _flutter =
+          FlutterRunTestDriver(new Directory('test/fixtures/flutter_app'));
+      await _flutter.run(withDebugger: true);
+      _flutterIsolateId = await _flutter.getFlutterIsolateId();
+
+      service = _flutter.vmService;
+    });
+
+    tearDown(() async {
+      await service.allFuturesCompleted.future;
+      await _flutter.stop();
+    });
+
+    /// Helper method to call an extension on the test device and verify that
+    /// the device reflects the new extension state.
+    Future<void> _enableExtensionOnTestDevice(
+      extensions.ToggleableServiceExtensionDescription extensionDescription,
+      Map<String, dynamic> params,
+      String evalExpression,
+      EvalOnDartLibrary library, {
+      String enabledValue,
+      String disabledValue,
+    }) async {
+      enabledValue ??= extensionDescription.enabledValue.toString();
+      disabledValue ??= extensionDescription.disabledValue.toString();
+
+      // Verify initial extension state on test device.
+      await _verifyExtensionStateOnTestDevice(
+        evalExpression,
+        disabledValue,
+        library,
+      );
+
+      // Enable service extension on test device.
+      await _flutter.callServiceExtension(
+        extensionDescription.extension,
+        params,
+      );
+
+      // Verify extension state after calling the service extension.
+      await _verifyExtensionStateOnTestDevice(
+        evalExpression,
+        enabledValue,
+        library,
+      );
+    }
+
+    /// Helper method to enable an extension on the test device, open the
+    /// vmService, and verify the enabled extension state is reflected by
+    /// [ServiceExtensionManager].
+    Future<void> _enableExtensionAndOpenVmService(
+      extensions.ToggleableServiceExtensionDescription extensionDescription,
+      Map<String, dynamic> params,
+      String evalExpression,
+      EvalOnDartLibrary library, {
+      String enabledValue,
+      String disabledValue,
+    }) async {
+      await _enableExtensionOnTestDevice(
+        extensionDescription,
+        params,
+        evalExpression,
+        library,
+        enabledValue: enabledValue,
+        disabledValue: disabledValue,
+      );
+
+      await serviceManager.vmServiceOpened(service, new Completer().future);
+
+      // Short delay for vmService to update extension states.
```

Comment:
you could also add a future on vmServiceManager that completes when all states have been updated.

Reply:
done.